### PR TITLE
Enable CL tools to read files without eigenvectors

### PR DESCRIPTION
**Pull request by rebeccafair**
I was playing with some old Phonopy data and it was annoying you couldn't plot `bands.yaml` without eigenvectors for `euphonic-dispersion` - now you can! I also updated `euphonic-dos` and `euphonic-intensity-map` - they now only need eigenvectors if you want to plot  PDOS or INS intensities.

### CODE REVIEW

**Comment by ajjackson** [comment]
This sounds very useful! Will take a look.

**Review by ajjackson — COMMENTED** [review]
This looks very handy!

I'm not quite convinced we've covered all the cases where a user loads frequencies and then tries to do eigenvector things with them, though.

I have this traceback:
```
(euphonic) ajj-mbp-1:~/src/euphonic (cl_no_evecs)$ euphonic-dos tests_and_analysis/test/data/qpoint_frequencies/quartz/quartz_bandstructure_qpoint_frequencies.json --pdos
Traceback (most recent call last):
  File "/Users/adam/mambaforge/envs/euphonic/bin/euphonic-dos", line 33, in <module>
    sys.exit(load_entry_point('euphonic', 'console_scripts', 'euphonic-dos')())
  File "/Users/adam/src/euphonic/euphonic/cli/dos.py", line 22, in main
    data = load_data_from_file(args.filename, frequencies_only)
  File "/Users/adam/src/euphonic/euphonic/cli/utils.py", line 149, in load_data_from_file
    return _load_json(path, frequencies_only)
  File "/Users/adam/src/euphonic/euphonic/cli/utils.py", line 114, in _load_json
    return modes_from_file(filename, frequencies_only)
  File "/Users/adam/src/euphonic/euphonic/cli/utils.py", line 95, in modes_from_file
    return cls.from_json_file(path)
  File "/Users/adam/src/euphonic/euphonic/qpoint_phonon_modes.py", line 582, in from_json_file
    return _obj_from_json_file(cls, filename,
  File "/Users/adam/src/euphonic/euphonic/io.py", line 119, in _obj_from_json_file
    return cls.from_dict(obj_dict)
  File "/Users/adam/src/euphonic/euphonic/qpoint_phonon_modes.py", line 569, in from_dict
    d['eigenvectors'], d['weights'])
KeyError: 'eigenvectors'
```

which suggests that somehow it is still ending up as a QpointPhononModes and then blowing up when eigenvalues are accessed.

**Comment by rebeccafair** [comment]
> This looks very handy!
> 
> I'm not quite convinced we've covered all the cases where a user loads frequencies and then tries to do eigenvector things with them, though.
> 
> I have this traceback:
> 
> ```
> (euphonic) ajj-mbp-1:~/src/euphonic (cl_no_evecs)$ euphonic-dos tests_and_analysis/test/data/qpoint_frequencies/quartz/quartz_bandstructure_qpoint_frequencies.json --pdos
> Traceback (most recent call last):
>   File "/Users/adam/mambaforge/envs/euphonic/bin/euphonic-dos", line 33, in <module>
>     sys.exit(load_entry_point('euphonic', 'console_scripts', 'euphonic-dos')())
>   File "/Users/adam/src/euphonic/euphonic/cli/dos.py", line 22, in main
>     data = load_data_from_file(args.filename, frequencies_only)
>   File "/Users/adam/src/euphonic/euphonic/cli/utils.py", line 149, in load_data_from_file
>     return _load_json(path, frequencies_only)
>   File "/Users/adam/src/euphonic/euphonic/cli/utils.py", line 114, in _load_json
>     return modes_from_file(filename, frequencies_only)
>   File "/Users/adam/src/euphonic/euphonic/cli/utils.py", line 95, in modes_from_file
>     return cls.from_json_file(path)
>   File "/Users/adam/src/euphonic/euphonic/qpoint_phonon_modes.py", line 582, in from_json_file
>     return _obj_from_json_file(cls, filename,
>   File "/Users/adam/src/euphonic/euphonic/io.py", line 119, in _obj_from_json_file
>     return cls.from_dict(obj_dict)
>   File "/Users/adam/src/euphonic/euphonic/qpoint_phonon_modes.py", line 569, in from_dict
>     d['eigenvectors'], d['weights'])
> KeyError: 'eigenvectors'
> ```
> 
> which suggests that somehow it is still ending up as a QpointPhononModes and then blowing up when eigenvalues are accessed.

To me this makes sense, you're calling with `--pdos`, which means surely you have the eigenvectors, right? 

I guess there are two ways to handle it:
1. Look at what's in the file and figure out whether it's `QpointFrequencies` or `QpointPhononModes` and load appropriately. Then complain if the user has set other arguments which are incompatible.
2. Look at the arguments and figure out what object is required, `QpointFrequencies` or `QpointPhononModes`, try and create this kind of object and leave Euphonic's file reading logic to throw whatever it throws, which is what's happened in this case.

I've gone with `2.` but looking at it now I'm thinking you had `1.` in mind when you refactored the command line tools (based on what's in `intensity_map.py`). I guess doing it that way could result with more helpful error messages. Maybe this needs a refactor already...

File: euphonic/cli/dispersion.py [hunk 1]
```diff
@@ -15,30 +15,32 @@
...
 def main(params: Optional[List[str]] = None) -> None:
     args = get_args(get_parser(), params)
-    data = load_data_from_file(args.filename)
 
-    if isinstance(data, euphonic.ForceConstants):
-        frequencies_only = not args.reorder  # Need eigenvectors to reorder
+    frequencies_only = not args.reorder  # Need eigenvectors to reorder
+
 
+    data = load_data_from_file(args.filename, frequencies_only)
```
[the remark this answers is not in the document]
Comment (rebeccafair) [reply euphonic/cli/dispersion.py]: Done

File: euphonic/cli/dos.py [hunk 2]
```diff
@@ -46,7 +47,7 @@ def main(params: Optional[List[str]] = None) -> None:
             modes = data.calculate_qpoint_phonon_modes(
                 mp_grid(grid_spec), **_calc_modes_kwargs(args))
 
-    elif isinstance(data, QpointPhononModes):
+    else:
```
[the remark this answers is not in the document]
Comment (rebeccafair) [reply euphonic/cli/dos.py:59]: I've refactored load_data_from_file and the CL tools to raise a ValueError with appropriate message if this happens, and added some tests for it.

File: tests_and_analysis/test/script_tests/test_dos.py [hunk 3]
```diff
@@ -37,7 +40,8 @@
     [quartz_fc_file, '--grid', '5', '5', '4'],
     [quartz_fc_file, '--grid', '5', '5', '4', '--adaptive', '--pdos'],
     [quartz_fc_file, '--grid', '5', '5', '4', '--adaptive'],
-    [quartz_fc_file, '--grid', '5', '5', '4', '--adaptive', '--eb', '2']]
+    [quartz_fc_file, '--grid', '5', '5', '4', '--adaptive', '--eb', '2'],
+    [nacl_no_evec_yaml_file]]
```
[the remark this answers is not in the document]
Comment (rebeccafair) [reply tests_and_analysis/test/script_tests/test_dos.py:44]: Agree, done

**Review by ajjackson — COMMENTED** [review]
Couple of tiny comments. Mostly visual inspection looks good! It all seems pretty clean and logical. I'll have a look at the test and try running a few things before approving.

File: euphonic/cli/dispersion.py [hunk 4]
```diff
@@ -15,30 +14,32 @@
...
-    data = load_data_from_file(args.filename)
 
-    if isinstance(data, euphonic.ForceConstants):
-        frequencies_only = not args.reorder  # Need eigenvectors to reorder
-
-        print("Force Constants data was loaded. Getting band path...")
+    frequencies_only = not args.reorder  # Need eigenvectors to reorder
+    data = load_data_from_file(args.filename, verbose=True,
+                               frequencies_only=frequencies_only)
+    if not frequencies_only and type(data) is QpointFrequencies:
```
[the remark this answers is not in the document]
Comment (rebeccafair) [reply euphonic/cli/dispersion.py:21]: Usually yes, but in this case `QpointPhononModes` (with eigenvectors) inherits from `QpointFrequencies` (no eigenvectors), so `isinstance(data, QpointFrequencies)` on a `QpointPhononModes` object returns `True`, and would trigger the `'Eigenvectors are required to use "--reorder" option'` error, not really what we want. So we can either check for an `eigenvectors` attribute, or specifically check the type. Maybe I should add a comment?

File: euphonic/cli/utils.py [hunk 5]
```diff
@@ -20,20 +20,74 @@
...
+def modes_from_file(filename: Union[str, os.PathLike]
+                    ) -> Union[QpointPhononModes, QpointFrequencies]:
+    warnings.warn('modes_from_file has been deprecated '
+                  'and will be removed in a future release. Please '
+                  'use load_data_from_file instead',
+                  category=DeprecationWarning,
+                  stacklevel=2)
+    data = load_data_from_file(filename)
+    if not isinstance(data, QpointFrequencies):
+        raise TypeError('File does not contain phonon modes')
```
[the remark this answers is not in the document]
Comment (rebeccafair) [reply euphonic/cli/utils.py:43]: Yeah, we can tackle this if it comes to it.

File: euphonic/cli/dispersion.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ajjackson) [reply euphonic/cli/dispersion.py:21]: I see!

It's a bit verbose, but

```python
if (not frequencies_only
    and isinstance(data, QpointFrequencies)
    and not isinstance(data, QpointPhononModes)
   ):
```
would be a precise option. Could avoid the `isinstance(data, QpointFrequencies)` by placing as an `elif` after the check for ForceConstants.

Or just leave it as-is until a real-world problem emerges.

File: tests_and_analysis/test/script_tests/test_dos.py [hunk 6]
```diff
@@ -23,6 +24,8 @@
     pass
 
 nah_phonon_file = get_castep_path('NaH', 'NaH.phonon')
+nacl_no_evec_yaml_file = os.path.join(
+    get_phonopy_path('NaCl', 'mesh'), 'mesh_no_evec.yaml')
```
Comment (ajjackson) [inline tests_and_analysis/test/script_tests/test_dos.py:28]: ### Not a request for this PR

Would be nice if these *get_path* functions returned Path objects, then these `join` calls could become e.g.

```python
nacl_no_evec_yaml_file = get_phonopy_path('NaCl', 'mesh') / 'mesh_no_evec.yaml'
```

Comment (ajjackson) [reply tests_and_analysis/test/script_tests/test_dos.py:28]: Or we should just let `get_phonopy_path` take multiple filenames? That would be closer to what it's supposed to do already.

```python
def get_phonopy_path(material: str, *filename: Tuple[str]) -> str:
    return os.path.join(get_data_path(), 'phonopy_files', material, *filename)
    
nacl_no_evec_yaml_file = get_phonopy_path('NaCl', 'mesh', 'mesh_no_evec.yaml')
```

**Review by ajjackson — APPROVED** [review]
Tests look thorough and systematic to me.

Have had a play with some data from Phonopy and it all seems to behave as expected.
(Well there is one thing actually, but it's not a regression so I'll open a new issue.)

All good! This makes for a nice kit of tools for playing with existing data.

**Pull request merged**